### PR TITLE
Migrate StateInput, CountryInput and custom fields to native `<select>` 

### LINKED ISSUE #44653: Google Autofill Causes State and Country Combo Boxes to Overlap on WooCommerce Checkout Page
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When using the Google Autofill feature to populate information on the checkout page that utilizes WooCommerce Checkout Blocks, the state and country combo boxes remain visible and overlap each other. This issue persists even after the autofill has completed, leading to a confusing and potentially error-prone user experience. The overlap makes it difficult for users to manually correct, verify,  or change the autofilled information.

### Expected behavior

Combo box should be correctly selected, but the drop down menu should not be present.

### Actual behavior

Combo boxes after autofill show the drop down menu and overlap eachother.

### Steps to reproduce

1. Go to my new test store: http://44.204.132.116/
2. Put the t-shirt in the cart and go to the checkout page.
3. Use Google Chrome Autofill to enter your details.

image

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): http://44.204.132.116
Site address (URL): http://44.204.132.116
WC Version: 8.5.2
REST API Version: ✔ 8.5.2
WC Blocks Version: ✔ 11.8.0-dev
Action Scheduler Version: ✔ 3.7.1
Log Directory Writable: ✔
WP Version: 6.4.3
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache
PHP Version: 8.2.15
PHP Post Max Size: 80 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.74.0
OpenSSL/1.1.1w

SUHOSIN Installed: –
MySQL Version: 11.2.2-MariaDB
Max Upload Size: 80 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.5.2
WC Database Prefix: wp_
Total Database Size: 5.26MB
Database Data Size: 3.56MB
Database Index Size: 1.70MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 2.47MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.06MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 2
page: 7
post: 2
product: 1
shop_order_placehold: 1
wp_navigation: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (2) ###

Jetpack: by Automattic – 12.9.3
WooCommerce: by Automattic – 8.5.2

### Inactive Plugins (9) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3.1
All-in-One WP Migration: by ServMask – 7.79
All in One SEO: by All in One SEO Team – 4.5.5
AMP: by AMP Project Contributors – 2.5.2
Google Analytics for WordPress by MonsterInsights: by MonsterInsights – 8.23.1
Hello Dolly: by Matt Mullenweg – 1.7.2
TaxoPress: by TaxoPress – 3.11.1
W3 Total Cache: by BoldGrid – 2.6.1
WP Mail SMTP: by WP Mail SMTP – 3.11.1

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to Woo.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/
Checkout: #9 - /checkout/
My account: #10 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Four
Version: 1.0
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-linked
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2024-02-16 17:01:26 +00:00
Options: ✔
Notes: 65
Onboarding: completed

### Action Scheduler ###

Complete: 7
Oldest: 2024-02-15 17:02:33 +0000
Newest: 2024-02-15 17:13:37 +0000

Failed: 1
Oldest: 2024-02-15 17:03:57 +0000
Newest: 2024-02-15 17:03:57 +0000

Pending: 1
Oldest: 2024-02-16 17:02:33 +0000
Newest: 2024-02-16 17:02:33 +0000

### Status report information ###

Generated at: 2024-02-15 17:13:58 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### LINKED ISSUE #47757: Replace Country/State and custom field inputs with native select
This issue is a blocker to https://github.com/woocommerce/woocommerce/issues/46357

Currently, these fields are rendered as a combobox, however it is unsuitable and keeps us on an old version of `@wordpress/components`.

The main reason we feel the combobox is not suitable for these fields is its unpredictability when it comes to handling user input and autocomplete. Address autocompletion is a very important thing to get right in the checkout flow and having a degraded experience is likely to harm conversion rates. The main issues with autocomplete are:

- Chrome autocomplete works differently to 1Password autocomplete
- There is no good way to check if a browser autocomplete is being used or if the user was typing
- When using autocomplete, this bug (https://github.com/woocommerce/woocommerce/issues/44653) occurs where United States addresses keep the dropdown list open, this is because there are two matches for United States (United States (US) and United States (US) Minor Outlying Islands)

A more suitable and accessible alternative is simply rendering a native `select` input for these fields.

### PR DESCRIPTION
**Pull request by samueljseay**
### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/47757
Closes https://github.com/woocommerce/woocommerce/issues/44653

This migrates the use of `@wordpress/components` Combobox to a native select instead. The rationale is that the Combobox does not work well for autofill and also by moving to a native select it will be more performant. 

Even without testing runtime performance this shrinks the bundle size by 40KB total while also removing the dependency on `lodash` in checkout, saving an additional 46KB.

### How to test the changes in this Pull Request:

There are 4 usages to test:

1. State Input in Billing and Shipping forms 
2. Country Input in Billing and Shipping forms
3. Custom fields as were implemented in https://github.com/woocommerce/woocommerce/issues/42758 

To test custom fields, you'll need a snippet like:

```php
add_action(
	'woocommerce_blocks_loaded',
	function() {
		woocommerce_register_additional_checkout_field(
			array(
				'id'             => 'plugin-namespace/road-type',
				'label'          => __( 'Road type', 'woo-gutenberg-products-block' ),
				'optionalLabel'  => __( 'Road type (optional)', 'woo-gutenberg-products-block' ),
				'autocomplete'   => 'road-type',
				'required'       => 'true',
				'autocapitalize' => 'characters',
				'location'       => 'address',
				'type'           => 'select',
				'options'        => array(
					array(
						'value' => 'alleyway',
						'label' => __( 'Alleyway', 'woo-gutenberg-products-block' ),
					),
					array(
						'value' => 'avenue',
						'label' => __( 'Avenue', 'woo-gutenberg-products-block' ),
					),
					array(
						'value' => 'boulevard',
						'label' => __( 'Boulevard', 'woo-gutenberg-products-block' ),
					),
					array(
						'value' => 'cul-de-sac',
						'label' => __( 'Cul-de-sac', 'woo-gutenberg-products-block' ),
					),
					array(
						'value' => 'street',
						'label' => __( 'Street', 'woo-gutenberg-products-block' ),
					),
				),
			)
		);
	}
);
```

4. Shipping calculator:

- Go to `WooCommerce -> Settings -> Shipping -> Shipping Settings` and check the `Enable the shipping calculator on the cart page` option.
- On the front-end, add an item to your cart and go to the Cart block.
- In the sidebar, under "Shipping" click "Change address"
- Use the country input to select various countries, ensure for countries with States (e.g. USA) the state field shows and functions correctly too.

5.  Keyboard navigation smoke testing

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Migrate the cart and checkout block's state, country and custom field input to a native select

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Comment by samueljseay** [comment]
Hey @opr @senadir @ralucaStan @nielslange while there is still a little bit of work to do and a question in the air about validation, I really need someone to start reviewing the implementation so we can catch any early problems. Most of the work is done.

Right now I will work on fixing the e2e tests in the meantime.

I would especially love if you could do some manual testing of look/feel/behaviour and provide feedback

**Review by opr — COMMENTED** [review]
Hi @samueljseay - looks like some extra commits got brought in here 👀  therefore I didn't do a code review, however I did check out the functionality and it seems great!

The list feels a little long when all the countries are displayed, but I don't know if we can shorten it to show fewer options at once?
image

I also noticed a small styling issue after autofilling, there's a small blank space at the bottom of the input

image

please re-request review once the merge has been fixed!

**Comment by samueljseay** [comment]
@opr Yes, sorry about that, this branch is a branch from the cart checkout bundle work so when I updated that branch wth trunk this one got out of date. fixed now. Thanks for the review, I'll explore what you've mentioned here.

**Comment by opr** [comment]
Noting that this PR also closes https://github.com/woocommerce/woocommerce/issues/44653 - I've updated the PR body.

**Review by nielslange — COMMENTED** [review]
~Thanks for working on this, @samueljseay. I've successfully tested this PR. The code changes look good as well, but it would be great to get another review from @opr and a review from @senadir.~

While checking the CI, I noticed a few issues. I'll re-review the PR again.

File: plugins/woocommerce-blocks/assets/js/base/components/select/style.scss
Comment (nielslange) [inline plugins/woocommerce-blocks/assets/js/base/components/select/style.scss]: This definition is a duplicate of line 5.

File: plugins/woocommerce-blocks/assets/js/base/components/select/style.scss
Comment (nielslange) [inline plugins/woocommerce-blocks/assets/js/base/components/select/style.scss:67]: This `transform` statement overwrites the one from line 47.

**Review by nielslange — CHANGES_REQUESTED** [review]
I noticed two CSS lint errors and one of the JS unit tests is failing as it still checks against `Combobox`. I also noticed that some blocks e2e tests failed running. I'll trigger them again now.

**Review by senadir — DISMISSED** [review]
I did a quick test for this and styles seems to be broken slightly, the outline is smaller than the actual button, and the select starts a bit too high.

I'm using 2024 with a small font size.

image

File: plugins/woocommerce-blocks/assets/js/base/components/select/style.scss
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/select/style.scss:67]: Thanks, that's what I get for copy paste 😅

**Comment by samueljseay** [comment]
@senadir apologies for the styling regression that slipped in there. I've addressed that now

File: plugins/woocommerce-blocks/assets/js/base/components/select/style.scss
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/select/style.scss]: Thanks for spotting this!

**Comment by samueljseay** [comment]
@nielslange @senadir this is ready for re-review. Styling issue fixed and most of the additional changes are refactoring of e2e tests

**Review by nielslange — CHANGES_REQUESTED** [review]
I noticed that the global styles are no longer applied to the native `<select>` fields:

<table>
<tr>
<td valign="top">Cart block:
<br><br>
Screenshot 2024-07-01 at 16 10 36
</td>
<td valign="top">Checkout block:
<br><br>
Screenshot 2024-07-01 at 16 11 01
</td>
</tr>
</table>

The screens above show the Rust styles of the Twenty Twenty-Four theme. I expect the following elements to appear in red:

- Text color, e.g. Germany
- Border color on focus
- Chevron color

**Comment by samueljseay** [comment]
Thanks @nielslange ! thats great spotting. will fix.

**Comment by samueljseay** [comment]
@nielslange I've addressed the theming issue in https://github.com/woocommerce/woocommerce/pull/48180/commits/180834279d7736c5e3918b1c3f2b4f62d545d537

**Review by nielslange — APPROVED** [review]
Thanks for addressing my previous feedback, @samueljseay. I've tested the PR successfully (both against Chrome, Firefox, Opera, and Safari in desktop and response mode) and found no issues in the code. Unless @opr or @senadir spot anything, I'd say let's ⛴️ this change.

**Review by senadir — COMMENTED** [review]
All is testing fine, one thing that seem to have regressed is that State is no longer 50% (or respecting that style, or getting).

In trunk, State is a half field, in this branch, it's always 100%.

Screenshot 2024-07-02 at 4 49 03 PM
Screenshot 2024-07-02 at 4 47 51 PM

**Comment by samueljseay** [comment]
Thanks @senadir for spotting that. I have changed the select to apply passed down classname to the top level element of the select component instead to fix this, this will work with the existing class better because it applies flex and half width to the component. I tested with this change and it now looks correct to me. Would you mind taking another look? 🙇🏻 Thanks for your time reviewing this.

**Pull request merged**
